multiuse-invites: Add edit feature for multiuse invites

bigBrain1901 opened the pull request:
> I am a GSoC '23 applicant, and this is my first PR :rocket: 

This pull request adds the feature to allow the editing of multiuse invites.

Partially fixes: GH-22099
- [ ] Blocking PR: GH-24977

### Screenshots and Screencast

**Added an "Edit" button on `SETTINGS / INVITATIONS` modal.**
<details><summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/55971005/223925453-1eedcf18-753e-442b-8f4b-223ddecf25b8.png">
</details>

**On clicking the "Edit" button, the "Edit Invite to Users" modal opens up. The only editable fields are `invite_as` and `streams`. The link itself is not editable, or shown in this modal.**
<details><summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/55971005/223925849-641e2e51-1e5a-4fdf-a6ac-a4100e98be21.png">
</details>

**On successful edit operation.**
<details><summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/55971005/223926382-d3f12e75-643a-4ddd-8a5d-ec58548cd458.png">
</details>

### Screencast of the Feature
[screen-recorder-thu-mar-09-2023-10-42-32.webm](https://user-images.githubusercontent.com/55971005/223926672-ad84a0f4-5146-4474-9453-f6b0040798ab.webm)

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).
- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.
- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks, this looks like a nice start!

In the settings menu:

1. We should make "Resend" an icon too -- let's try https://fontawesome.com/v4/icon/paper-plane.
2. Let's use a trash can for the "revoke" icon, for consistency with other settings.
3. All the action buttons need tooltips: "Resend", "Edit", "Revoke invite" should work.

I also get an error whey I click the edit button on an email invite. I think the button shouldn't be there in the first place?
<img width="959" alt="Screen Shot 2023-03-13 at 11 13 55 PM" src="https://user-images.githubusercontent.com/2090066/224912203-808149c9-a6fe-41d1-8e37-f3a727c6d146.png">

alya commented:
In the invite modal:

1. If I change something and then undo (e.g., check and uncheck a stream), the "Save changes" button is active. Ideally, it would go back to being inactive, though it's not critical if it's hard to do.
2. Let's change the modal invite to just "Edit invite".

alya commented:
Oh,  it looks like your changes have affected other "edit" buttons?

<img width="945" alt="Screen Shot 2023-03-13 at 11 20 37 PM" src="https://user-images.githubusercontent.com/2090066/224913214-045662dc-8058-49e8-b7e4-0781c323792b.png">

alya commented:
Once all the feedback has been addressed, please post updated screenshots and request another round of review. Please also do a thorough round of manual testing to make sure you're not seeing errors.

alya commented:
Thanks!

> Any change to the selection of streams now enables or disables the "Save changes" button based on - at least 1 stream must be selected.

Ah, I think there was some misunderstanding. Not having any streams selected is totally valid. I was noting that if you select and then unselect the same stream (or unselect and then select), so that no actual changes are made, the button is still active.

alya commented:
> I have added a single commit with the round 1 review changes.

Please squash all the changes you are making into your initial commit.

bigBrain1901 commented:
> Ah, I think there was some misunderstanding. Not having any streams selected is totally valid, and the user should be able to save this configuration. I was noting that if you select and then unselect the same stream (or unselect and then select), so that no actual changes are made, the button is still active.

Oh alright! I wasn't sure about this assumption I made.

I have added my implementation path in the PR description (towards the end). There may be some corrections I need there as well, in case I misunderstood some aspect of this issue. Please check that as well.

alya commented:
I'm not reviewing the implementation (that will be a separate step), but as I said above, this is incorrect behavior:

> Assumed a policy from code - At least 1 stream has to be selected for every multiuse invite. In other words, a user cannot edit an invite to set multiuse_invite.streams = [ ]. This might be invalid, needs a check.

bigBrain1901 commented:
Okay, understood, thanks a lot for the info!

> Ah, I think there was some misunderstanding. Not having any streams selected is totally valid, and the user should be able to save this configuration. I was noting that if you select and then unselect the same stream (or unselect and then select), so that no actual changes are made, the button is still active.

This misunderstanding has now been cleared and fixed in code. The PR is ready for review again.

alya commented:
Thanks!

If I click into a modal and then exit out of it, the tooltips incorrectly stick around. E.g., clicking on "Revoke invite" and then hovering over "Resend":

![Screen Shot 2023-03-14 at 1 06 28 PM](https://user-images.githubusercontent.com/2090066/225124608-61b49de2-5b60-463a-ae92-0a027ddc38e2.png)

alya commented:
> If I change something and then undo (e.g., check and uncheck a stream), the "Save changes" button is active. 

This part looks to be fixed for the checkboxes, but not the role dropdown.

alya commented:
Playing around with it some more, let's change it so that:

1. If the changes to the invite succeed, the modal is closed. There's no confirmation banner. Otherwise, it's confusing what you're supposed to do next.
2. If somehow there's an error saving changes, the modal remains open, and an error banner is displayed.

This matches the behavior of the "Manage user" modal.

alya commented:
> I was unable to reproduce the bug. Here is a creencast.

Hm, OK, I'm not seeing that anymore either. :)

alya commented:
@sahil839 could you please take a look at this PR? It sounds like it should be reviewable despite the failing check.

> The `Install production` step failed for `Ubuntu 22.04` on the previous push due to a connection timeout. Curious to know what went wrong here 👼

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -213,7 +296,85 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+        e.preventDefault();
+
+        const invite_id = Number.parseInt(e.currentTarget.dataset.inviteId, 10);
+        const invite = meta.invites.find((invite) => invite.id === invite_id && invite.is_multiuse);
```
Though this works, but we can maybe do it by attaching a `data-invite-id` attribute to the row element. We also do this for the users and bots list, so would be nice to have some consistency.

sahil839 commented on web/templates/settings/edit_invite_user_modal.hbs:
```diff
@@ -0,0 +1,50 @@
+<form id="invite-user-form">{{ csrf_input }}
+    {{#if development_environment}}
+        <div class="alert" id="dev_env_msg"></div>
```
Do we need this for multiuse-invites as well? I think it is only used to show that users can check the emails in development environment at `/emails`.

bigBrain1901 replied:
Ah yes! This is not needed for multiuse-invites.

sahil839 reviewed (commented):
Thanks for working on this. Posted a batch of comments. We would ideally want to have all the preparatory work and changes other than to allow edits to multi-use invites be in separate commits. See https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html for more details.

bigBrain1901 commented on zerver/tests/test_events.py:
```diff
@@ -860,6 +860,167 @@ def test_revoke_user_invite_event(self) -> None:
         )
         check_invites_changed("events[0]", events[0])
 
+    def test_multiuse_invite_edits_with_malformed_parameters(self) -> None:
```
Hi, I have a doubt.

Would it be okay to use the classdef `class MultiuseInviteActionsTest(BaseAction):` for the new test file?

`BaseAction` is coming from `test_actions.py`, useful for running `self.verify_action(...)` in the new test-file.

bigBrain1901 commented on web/src/settings_invites.js:
```diff
@@ -177,13 +186,87 @@ export function set_up(initialize_event_handlers = true) {
+                    has_billing_access: page_params.is_owner || page_params.is_billing_admin,
+                    daily_limit_reached: arr.daily_limit_reached,
+                });
+                ui_report.message(error_response, $invite_status, "alert-warning");
```
Yes. Just sticking to a simple ui_report.warning_message will do the job.

```js
error(xhr) {
    const arr = JSON.parse(xhr.responseText);
    ui_report.message(arr.msg, $invite_status, "alert-warning");
},
```

Does this look fine? Please lmk.

bigBrain1901 commented on web/templates/settings/edit_invite_user_modal.hbs:
```diff
@@ -0,0 +1,50 @@
+                        {{/if}}
+                    </label>
+                {{/each}}
+            </div>
```
Please check if the templating is fixed as desired.

bigBrain1901 commented:
Thanks for the review! I have made the required changes and left a few doubts in the comments.

The PR is ready for review again.

bigBrain1901 commented:
@sahil839 
Requesting for review. Please lmk whenever free 🚀

sahil839 commented on web/src/invite.js:
```diff
@@ -256,6 +257,9 @@ function open_invite_user_modal(e) {
         $("#invite-user-modal .dialog_submit_button").prop("disabled", true);
+        for (const stream of streams.filter((stream) => stream.default_stream)) {
+            $(`input[value=${stream.stream_id}]`).prop("checked", true);
+        }
```
This should be moved to the commit which changes the template code to not handle marking the default streams initially and the reason for this change should be mentioned in the commit message

sahil839 commented on zerver/tests/test_events.py:
```diff
@@ -2609,6 +2610,35 @@ def test_notification_setting_event_not_sent(self) -> None:
+            ),
+            state_change_expected=False,
+            event_types=["invites_changed"],
+        )
```
You should add a call to `check_invites_changed` to verify the contents of the event.

bigBrain1901 replied:
```py
self.verify_action(
    lambda: do_edit_multiuse_invite_link(
        initial_multiuse_object, initial_invited_as, initial_streams
    ),
    state_change_expected=False,
    event_types=["invites_changed"],
)
```

I inferred that passing `event_types=[...]` to `verify_action` checked the event type for me. Now that I see the test file, the helper `check_invites_changed` is used consistently.

I want to know if consistency is the reason for this review comment or was my inference very off. Please lmk.

sahil839 replied:
`event_types` is useful when we want to test only events of a particular type where multiple type events are sent. To check the event object, which includes type field, we use the `check_..` functions.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -214,7 +263,85 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+        });
+    });
+
+    $(".admin_invites_table").on("click", ".resend-invite", (e) => {
```
This change should be in the commit which changes the class name.

sahil839 commented on web/templates/invite_as_and_streams_to_join.hbs:
I think we can have a better file name. I can think of `invite_role_and_streams_info` right now, but there can be something better.

bigBrain1901 replied:
Maybe `invite_role_and_streams_form` is a better name as this template is just a partial form with those 2 elements?  
Or just `invite_role_and_streams` because I cannot see any `_form` partials?

sahil839 replied:
I don't have any strong opinions. Anyways, we can also leave it for now and update if needed after feedback from others.

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -382,6 +382,7 @@ def do_get_invites_controlled_by_user(user_profile: UserProfile) -> List[Dict[st
                 ),
                 invited_as=invite.invited_as,
                 is_multiuse=True,
+                streams=list(invite.streams.values_list("id", flat=True)),
```
I think this change can be moved to a separate preparatory commit and also you should add corresponding code in tests to check this field.

bigBrain1901 replied:
This is what I propose to do for the test.  
Since there is no existing test for checking the contents of the invite objects, we can add a test as shown below in `zerver.tests.test_invite.InvitationsTestCase`

```py
def test_content_of_invite_objects(self) -> None:
        """
        Checks content of the invite objects returned by do_get_invites_controlled_by_user.
        """
        # there will be 2 things tested here
        ## 1. test for all fields in a email-based user invite
        ## 1. test for all fields in a multiuse invite
```

Does this sound okay?  
I will place these tests in appropriate commits.

bigBrain1901 replied:
This is what I propose to do for the test.  
Since there is no existing test for checking the contents of the invite objects, we can add a test as shown below in `zerver.tests.test_invite.InvitationsTestCase`

```py
def test_content_of_invite_objects(self) -> None:
        """
        Checks content of the invite objects returned by do_get_invites_controlled_by_user.
        """
        # there will be 2 things tested here
        ## 1. test for all fields in a email-based user invite
        ## 1. test for all fields in a multiuse invite
```

Does this sound okay?  
I will place these tests in appropriate commits.

bigBrain1901 replied:
Please verify if the test files are structured as desired.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -214,7 +263,85 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+                $("#invite-user-modal .dialog_submit_button").prop("disabled", false);
+            }
+            function disable_invite_submit_button() {
+                $("#invite-user-modal .dialog_submit_button").prop("disabled", true);
```
I don't think we need these functions as they are just a single line of code.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -214,7 +263,85 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+            }
+
+            // Checks the state of the modal and enables or disables the save changes button
+            function toggle_submit_button() {
```
Adding this function is probably overkill. You can just call `$("#invite-user-modal .dialog_submit_button").prop("disabled", state_changed());`.

sahil839 reviewed (commented):
Posted some comments. A few points regarding commit structure -
- We ideally prefer to keep unrelated changes in different small commits since it makes it easy to review and merge them incrementally. 
- For this PR, you can have different commits - one for extracting the common template file, one for renaming the class of buttons for adding the icons and tooltips to buttons, and one for adding `streams` field to the data sent by server.
- Also, the commit messages for this commit should clearly mention what and why the changes have been done instead of just mentioning that it is a preparatory commit. Please read the [commit message guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages) for more details.

sahil839 commented:
Though we do not document the invite related endpoints currently, but we still have entries for them in `changelog.md`, so would be good to add it for these too. See point 11 in https://zulip.readthedocs.io/en/latest/documentation/api.html#step-by-step-guide, we can probably skip adding documentation for this endpoint in `zulip.yaml` for now, but the rest regarding changes in `changelog.md` and `version.py` should be done.

sahil839 commented on zerver/views/invite.py:
```diff
@@ -156,6 +157,40 @@ def revoke_multiuse_invite(
+    if not user_profile.can_invite_others_to_realm():
+        # Guest users case will not be handled here as it will
+        # be handled by the decorator above.
+        raise JsonableError(_("Insufficient permission"))
```
I guess we do not need this since we anyways allow admins to create multi-use links always irrespective of this setting.

sahil839 replied:
@alya @timabbott I got one more question setting this, do we want to allow creating multi-use links when `Who can invite users to this organization` setting is set to `Nobody`. Currently we allow it.

bigBrain1901 commented on zerver/views/invite.py:
```diff
@@ -156,6 +157,52 @@ def revoke_multiuse_invite(
+    except MultiuseInvite.DoesNotExist:
+        raise JsonableError(
+            _("Invite does not exist with id: {}. No invites were sent.").format(prereg_id)
+        )
```
Oops, apologies! I will be careful from now on.

bigBrain1901 commented:
Hi @sahil839, thanks for all the feedback :rocket: 
I have made the required changes.

Please verify if the changes are as desired. I am still not fully sure if I have the following correctly done -
1. Test files and their composition.
2. Commit messages.

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -2375,3 +2375,153 @@ def test_create_multiuse_link_invalid_invite_as_api_call(self) -> None:
+        For a successful edit to a multiuse_invite, it is essential
+        that the following situation occurs.
+        * The editing user must be permitted to invite others
+        * If the link invites users as moderators or realm admins,
```
And the test should be modified accordingly.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -178,6 +184,48 @@ export function set_up(initialize_event_handlers = true) {
+
+function submit_invitation_form(invite_id) {
+    const $invite_status = $("#dialog_error");
+    const data = get_invitation_data(invite_id);
```
There is no need to pass `invite_id` parameter here.

bigBrain1901 replied:
I used `meta.invite` to store the `multiuse_invite` that is being edited currently.

Is this approach alright?

sahil839 commented on zerver/views/invite.py:
```diff
@@ -156,6 +157,43 @@ def revoke_multiuse_invite(
+def edit_multiuse_invite(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    prereg_id: int,
```
This should be `invite_id`, since we use `MultiuseInvite` object for multi-use invites and not `PreregistrationUser` obejcts.

bigBrain1901 replied:
I shifted the endpoint for the update from `PATCH /json/invites/:id` to `PATCH /json/invites/multiuse/:id`, which I originally intended to use as well.

Please verify if this is ok.

sahil839 replied:
Yeah, looks good for now to me.

sahil839 commented on web/templates/settings/admin_invites_list.hbs:
```diff
@@ -33,12 +33,23 @@
+              class="button rounded small resend-invite btn-warning tippy-zulip-tooltip"
+              {{#if disable_buttons}}disabled="disabled"{{/if}}
+              data-invite-id="{{id}}"
+              data-tippy-content="{{t 'Resend'}}"
```
I am not sure whether this reformatting help in readability. Since, I am not sure, I would wait for feedback from others on this.

sahil839 reviewed (commented):
Posted some comments. A couple of other points -
- The tooltips for the button for the first invite get cut from top.
- The commit messages can still be improved. In first commit message summary can be `Replace the button text with icons` and also it should be mentioned that the class is being renamed for clarity. In the second message, the prefix `templating` should be changed to either `templates` or `invite`. In the third commit, it would be better to mention that change is only for multiuse invites.
- Also, there should be a changelog entry for the third commit as well.

bigBrain1901 commented:
> The tooltips for the button for the first invite get cut from top.

I was unable to reproduce this bug. Here is a screenshot of the invites list. Is this the correct button you are referring to?
![image](https://user-images.githubusercontent.com/55971005/227641398-e6458e99-582f-431a-8d28-23a52debe367.png)

bigBrain1901 commented:
@sahil839 thanks for the review again!

I have resolved all the comments and replied to some with a few doubts.  
Please take a look at these as well.

sahil839 commented on zerver/tests/test_invite.py:
```diff
@@ -1662,18 +1665,45 @@ def test_successful_get_open_invitations(self) -> None:
+        self.assertEqual(
+            invites[1]["expiry_date"] - invites[1]["invited"], invite_expires_in_minutes * 60
+        )
+        self.assertEqual(set(invites[1]["streams"]), set(stream_ids))
```
Can be done directly with a PR when needed after feedback from others.

bigBrain1901 commented:
> Regarding the tooltip bug, I think it is only reproducible if there is only one invite in the list ...

On the Bots and User settings modals, the tooltips used are not tippy, but standard HTML tooltips (using the title attribute).

1. Do you suggest that I use the same format?
2. Should I write a fix in a new commit, that addresses the bug? Maybe, this can be a new issue in itself.

bigBrain1901 commented on zerver/views/invite.py:
```diff
@@ -156,6 +157,34 @@ def revoke_multiuse_invite(
+    except MultiuseInvite.DoesNotExist:
+        raise JsonableError(_("Invite does not exist with id: {}.").format(invite_id))
+
+    check_if_owner_required(invite_as, user_profile)
```
> I am not sure about one thing - do we want to allow admins to edit invites that originally add `invite_as` as `Owners`?

I see a use case where I would invite all `owners` of a set of streams with a multiuse link and then re-purpose the same link to invite other `members` to those streams. I mean, this flexibility can be given to the users, by code.

Is there some other policy preventing this from happening?

bigBrain1901 reviewed (commented):
Thanks for the feedback! Here are a few points to verify that I am not sure about.

bigBrain1901 commented on web/src/settings_invites.js:
```diff
@@ -215,6 +264,69 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+            is_owner: page_params.is_owner,
+            streams: get_invite_streams(),
+            notifications_stream: stream_data.get_notifications_stream(),
+            invite: meta.invite,
```
No, it is not required.

Pushed the change.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -178,6 +184,46 @@ export function set_up(initialize_event_handlers = true) {
+    return data;
+}
+
+function do_edit_invite() {
+    const $invite_status = $("#dialog_error");
+    const data = get_invitation_data();
+
+    channel.patch({
+        url: "/json/invites/multiuse/" + meta.invite_id,
+        data,
+        beforeSend,
+        success() {
+            dialog_widget.close_modal();
+        },
+        error(xhr) {
+            const arr = JSON.parse(xhr.responseText);
+            ui_report.message(arr.msg, $invite_status, "alert-warning");
+        },
+        complete() {
+            $("#edit-invite-form .dialog_submit_button").text($t({defaultMessage: "Save changes"}));
+            $("#edit-invite-form .dialog_submit_button").prop("disabled", false);
+            $("#edit-invite-form .dialog_cancel_button").prop("disabled", false);
+        },
+    });
+}
```
Yes, seems good.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -215,6 +261,70 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+        e.stopPropagation();
+        e.preventDefault();
+
+        meta.invite_id = Number.parseInt($(e.currentTarget).attr("data-invite-id"), 10);
```
I was thinking of passing it to `do_edit_invite` function as parameter. But this seems fine for now if we do the same for revoking invite.

sahil839 reviewed (commented):
Posted a couple of comments.

Regarding tooltips, I think using `data-tippy-placement` attribute should be fine and this change should be done in the commit which adds the tooltip. We are moving away from the html title attributes to consistently use tippy tooltips.

timabbott commented:
Does it make sense for @bigBrain1901 to create a new PR that's just the backend portion of this PR, so that we can try to integrate the backend changes without the frontend being a distraction? The review thread here is already very long. (@sahil839 I defer to you on the plan, since I've not read all the feedback above)

bigBrain1901 commented on web/src/settings_invites.js:
```diff
@@ -8,6 +8,7 @@ import * as blueslip from "./blueslip";
 import * as channel from "./channel";
 import * as confirm_dialog from "./confirm_dialog";
 import {$t, $t_html} from "./i18n";
+import {get_invite_streams} from "./invite";
```
In this commit (templates: Create a form partial ...), the `get_invite_streams` function is used in this file.

In the last commit (invites: Add the ability to edit ...), the function `get_invite_streams` is still being used, and this line is not removed per say, just edited from  
`import {get_invite_streams} from "./invite";;`  
to  
`import {beforeSend, get_invite_streams} from "./invite"`

bigBrain1901 commented on web/src/tippyjs.js:
```diff
@@ -468,4 +468,9 @@ export function initialize() {
+    delegate("body", {
+        target: ".revoke-invite, resend-invite",
+        appendTo: () => document.querySelector(".organization-box"),
+    });
```
This had to be added because the default behaviour for tippy tooltips in Zulip is to appendTo parent. However, that causes the tooltip to get cut when used within the settings/invitations modal's table.

`.organization-box` is the className of the content div of the settings modal.

bigBrain1901 replied:
<details>
<summary>Screenshot after the change</summary>
<img src="https://user-images.githubusercontent.com/55971005/229035478-756b9082-349e-45db-aa14-b936e9d3ee87.png">
</details>

bigBrain1901 replied:
Had opened [this thread on CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/.E2.9C.94.20tippy.20tooltip.20with.20tables) to discuss the same. Resolved now.

bigBrain1901 commented:
@sahil839, I have amended the tooltips for `settings/invitations`. Doing the same for other `settings/*` pages could be a new PR.

---

Meanwhile, here are the threads to revisit:

1. [Who is allowed to edit invites to owners?](https://github.com/zulip/zulip/pull/24666#discussion_r1149208158)
2. [Does this template formatting help with readability?](https://github.com/zulip/zulip/pull/24666#discussion_r1147515965)
3. [Should we allow creating/edit of multiuse invites when the `Who can invite users to this organization` is set to `NOBODY`](https://github.com/zulip/zulip/pull/24666#discussion_r1144620756)

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -214,6 +261,70 @@ export function on_load_success(invites_data, initialize_event_handlers) {
+            });
+            $("#invite_uncheck_all_button").on("click", () => {
+                $("#streams_to_add :checkbox").prop("checked", false);
+                $("#edit-invite-form .dialog_submit_button").prop("disabled", state_unchanged());
```
Just noting here that there is a PR opened recently to add this functionality of enabling/disabling the button for all modals. #24911 is the PR. Nothing to change for now, but we can update this if #24911 gets merged first.

sahil839 reviewed (commented):
Posted one comment. Also, I am not sure why, but I cannot see tolltip on the resend invite button for email invites.

bigBrain1901 commented:
> > Does it make sense for @bigBrain1901 to create a new PR that's just the backend portion of this PR, so that we can try to integrate the backend changes without the frontend being a distraction? The review thread here is already very long. (@sahil839 I defer to you on the plan, since I've not read all the feedback above)
> 
> I don't think I would have more feedback on the code now, but could still be helpful to have a separate PR for backend poriton for your review.

@sahil839 I have fixed the UI bug with the Revoke tooltip and tested the changes as well.  
Shall I go ahead and split the PR?

Also, these threads need more feedback - [Link to thread](https://github.com/zulip/zulip/pull/24666#issuecomment-1491321246)

sahil839 commented on web/src/tippyjs.js:
```diff
@@ -470,7 +470,7 @@ export function initialize() {
 
     delegate("body", {
-        target: ".revoke-invite, resend-invite",
+        target: ".edit-invite, .revoke-invite, .resend-invite",
```
The "resend-invite" -> ".resend-invite" change should be done in the commit which adds the tooltip. We try to include unrelated changes in the commit to maintain a clean history.

sahil839 commented on web/templates/settings/admin_invites_list.hbs:
```diff
@@ -41,13 +51,13 @@
           data-invite-id="{{id}}"
           data-is-multiuse="{{is_multiuse}}"
-          data-tippy-content="{{t 'Revoke invite'}}"
+          data-tippy-content="{{t 'Revoke'}}"
```
This change should also be in the commit which adds the tooltip as mentioned in the previous comment.

sahil839 reviewed (commented):
Posted a couple of comments.

Yes, you can open a separate PR for backend part. And also would be helpful if you could quote the pending backend related discussion in that PR with links pointing to the comments in this PR.

bigBrain1901 commented:
Hi! I have created the new GH-24977 and referred to the threads to be reviewed from this PR.  
Please lmk if I missed something :rocket:

timabbott commented:
Closing in favor of #32830, which continues this work.
